gateway-shell/src/main/java/org/apache/knox/gateway/shell/KnoxSh.java

```diff
@@ -155,6 +170,81 @@ public boolean validate() {
     public abstract String getUsage();
   }
 
+  private class KnoxBuildTrustStore extends Command {
+
+    private static final String USAGE = "buildTrustStore --gateway server-url";
+    private static final String DESC = "Downloads the gateway server's public certificate and builds a trust store.";
+    private static final String CLIENT_TRUST_STORE_FILE_NAME = "gateway-client-trust.jks";
+
+    @Override
+    public void execute() throws Exception {
+      final X509Certificate gatewayServerPublicCert = fetchPublicCertFromGatewayServer();
+      if (gatewayServerPublicCert != null) {
+        final File trustStoreFile = new File(System.getProperty("user.home"), CLIENT_TRUST_STORE_FILE_NAME);
+        X509CertificateUtil.writeCertificateToJks(gatewayServerPublicCert, trustStoreFile);
+        out.println("Gateway server's certificate is exported into " + trustStoreFile.getAbsolutePath());
+      } else {
+        out.println("Could not obtain server certificate chain");
+      }
+    }
+
+    private X509Certificate fetchPublicCertFromGatewayServer() throws Exception {
+      final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+      trustManagerFactory.init((KeyStore) null);
+      final X509TrustManager defaultTrustManager = (X509TrustManager) trustManagerFactory.getTrustManagers()[0];
+      final CertificateChainAwareTrustManager trustManagerWithCertificateChain = new CertificateChainAwareTrustManager(defaultTrustManager);
+      final SSLContext sslContext = SSLContext.getInstance("TLS");
+      sslContext.init(null, new TrustManager[] { trustManagerWithCertificateChain }, null);
+
+      final URI uri = URI.create(gateway);
+      out.println("Opening connection to " + uri.getHost() + ":" + uri.getPort() + "...");
+      final SSLSocket socket = (SSLSocket) sslContext.getSocketFactory().createSocket(uri.getHost(), uri.getPort());
+      socket.setSoTimeout(10000);
+      try {
+        out.println("Starting SSL handshake...");
+        socket.startHandshake();
+        socket.close();
+        out.println();
+        out.println("No errors, certificate is already trusted");
+      } catch (SSLException e) {
+        // NOP; this is expected in case the gateway server's certificate is not in the trust store the JVM uses
+        out.println("SSL exception; found non-trusted certificate");
+      }
+
+      return trustManagerWithCertificateChain.certificateChain == null ? null : trustManagerWithCertificateChain.certificateChain[0];
```

Comment:
Will this return just the one public certificate or the chain of trust? The reason the chain is important is because there could be multiple intermediate certificates.

Reply:
Or are we just going to directly trust the final server certificate?

Reply:
It's enough; right? We only need to have one certificate in the KnoxShell's truststore with 'gateway-identity` alias.

Reply:
I'm not sure. The Knox certificate could be externally configured or have a CA signed certificate in `gateway.jks` The server should present all the certificates. It would be good to test this with a locally CA signed certificate.

Reply:
I guess the question I have, is there any harm in returning the whole certificate chain?